addons/website_sale/controllers/main.py

```diff
@@ -226,19 +205,17 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
                 parent_category_ids.append(current_category.parent_id.id)
                 current_category = current_category.parent_id
 
-        product_count = product_obj.search_count(cr, uid, domain, context=context)
+        product_count = Product.search_count(domain)
         pager = request.website.pager(url=url, total=product_count, page=page, step=ppg, scope=7, url_args=post)
-        product_ids = product_obj.search(cr, uid, domain, limit=ppg, offset=pager['offset'], order='website_published desc, website_sequence desc', context=context)
-        products = product_obj.browse(cr, uid, product_ids, context=context)
+        products = Product.with_context(pricelist_context).search(domain, limit=ppg, offset=pager['offset'], order='website_published desc, website_sequence desc')
 
-        attributes_obj = request.registry['product.attribute']
-        if product_ids:
-            attributes_ids = attributes_obj.search(cr, uid, [('attribute_line_ids.product_tmpl_id', 'in', product_ids)], context=context)
-        attributes = attributes_obj.browse(cr, uid, attributes_ids, context=context)
+        ProductAttribute = request.env['product.attribute']
+        if products:
+            attributes = ProductAttribute.search([('attribute_line_ids.product_tmpl_id', 'in', products.ids)])
+        else:
+            attributes = ProductAttribute.browse(attributes_ids)
 
-        from_currency = pool['res.users'].browse(cr, uid, uid, context=context).company_id.currency_id
-        to_currency = pricelist.currency_id
-        compute_currency = lambda price: pool['res.currency']._compute(cr, uid, from_currency, to_currency, price, context=context)
+        compute_currency = lambda price: request.env.user.company_id.currency_id.compute(price, pricelist.currency_id)
```

Comment:
with your code, you will recompute the currency each time we execute the `compute_currency` method. There is cache but still...